sdk-cpp/src/do_download.cpp

```diff
@@ -58,23 +58,61 @@ void download::download_url_to_path(const std::string& uri, const std::string& d
     download_url_to_path(uri, downloadFilePath, isCancelled, timeOut);
 }
 
+// Ensures download is finalized/aborted in the agent to avoid stale download objects.
+// Future: We could expect callers to explicitly opt-in to persist downloads in the agent and
+//   use that as an indicator to perform/skip the cleanup directly in ~download().
+class download_cleanup_call
+{
+private:
+    download* _ptr { nullptr };
+
+public:
+    download_cleanup_call(download* pObj) :
+        _ptr(pObj)
+    {
+    }
+
+    ~download_cleanup_call()
+    {
+        if (_ptr != nullptr)
+        {
+            try
+            {
+                const auto state = _ptr->get_status().state();
+                if (state == download_state::transferred)
+                {
+                    _ptr->finalize();
+                }
+                else if ((state != download_state::aborted) && (state != download_state::finalized))
+                {
+                    _ptr->abort();
+                }
+            }
+            catch (...)
+            {
+            }
+        }
+    }
+};
+
 void download::download_url_to_path(const std::string& uri, const std::string& downloadFilePath, const std::atomic_bool& isCancelled, std::chrono::seconds timeOut)
 {
     constexpr std::chrono::seconds maxPollTime = 5s;
     std::chrono::milliseconds pollTime = 500ms;
     const auto endTime = std::chrono::system_clock::now() + timeOut;
 
     download oneShotDownload(uri, downloadFilePath);
+    download_cleanup_call scopedCleanup{&oneShotDownload};
+
     oneShotDownload.start();
```

Comment:
I noticed that a failure in start() was causing abort() on line 133 to not get called (exception gets thrown), leaving a stale download object in the agent.